rllib/evaluation/sampler.py

```diff
@@ -649,13 +649,31 @@ def new_episode(env_id):
                     episode=episode,
                     tf_sess=p.get_session(),
                 )
-        callbacks.on_episode_start(
-            worker=worker,
-            base_env=base_env,
-            policies=worker.policy_map,
-            episode=episode,
-            env_index=env_id,
-        )
+        if worker.policy_config["in_evaluation"]:
+            if getattr(callbacks, "on_evaluate_start") is not None:
```

Comment:
As a separate point, I have added these checks just in case of compatibility issues, but do you think we can remove them instead?

Reply:
hmm, this doesn't look right. you don't want on_evaluate_start() to be called before every single eval episode right?
you want it to be called once before the whole eval thing starts.
so I feel like you want to do it here:
https://github.com/ray-project/ray/blob/3dd9c18c94254bcd3655bb91cc98b643631a6dfa/rllib/algorithms/algorithm.py#L706

same for the on_evaluate_end() call.

Reply:
I wanted to have some per-episode metrics, at least for the on_evaluate_end(), and I don't know if there is a way to still gather that from within the evaluate function. 

I guess putting them in evaluate() would be mostly the same, unless evaluate length is more than 1?

Reply:
sorry about the delay. had a vacation, and then the ray summit.
coming back to this PR, my point is that if you want to get per-episode stats for evaluation episodes, you can do it simply in on_episode_end(). because you have access to worker.policy_config in the callback, you can tell whether that episode is for eval or not.
you do not need any additional extension points to achieve that.
feel like on_evaluate_start() and on_evaluate_end() should only be called when the entire evaluation is kicked off / finished?

Reply:
I hope you had a good time on your holidays, and at the ray summit! :)

I see your point, and I agree - I will move both calls to the algorithm. Likewise, I can also include the documentation in terms of accessing the evaluation episodes from on_episode_end() into this PR?

Reply:
wow, if you can include the documentation change in this PR, that'd be a huge help for the community!!
doc changes for the win! :)